Documentation/project-docs/AssemblyBlobs.md

```diff
@@ -0,0 +1,214 @@
+
+# Blob kinds and locations
+
+Each application will contain at least a single blob, with assemblies
+that are architecture-agnostics and any number of
+architecture-specific blobs.  dotnet ships with a handful of
+assemblies that **are** architecture-specific - those assemblies are
+placed in an architecture specific blob, one per architecture
+supported by and enabled for the application.  On the execution time,
+the Xamarin.Android runtime will always map the architecture-agnostic
+blob and one, and **only** one, of the architecture-specific blobs.
+
+Blobs are placed in the same location in the APK/AAB archive where the
+individual assemblies traditionally live, the `assemblies/` (for APK)
+and `base/root/assemblies/` (for AAB) folders.
+
+The architecture agnostic blob is always named `assemblies.blob` while
+the architecture-specific one is called `assemblies.[ARCH].blob`.
+
+Currently, Xamarin.Android applications will produce only one set of
+blobs but when Xamarin.Android adds support for Android Features, each
+feature APK will contain its own set of blobs.  All of the APKs will
+follow the location, format and naming conventions described above.
+
+# Blob format
+
+Each blob is a structured binary file, using little-endian byte order
+and aligned to a byte boundary.  Each blob consists of a header, an
+assembly descriptor table and, optionally (see below), two tables with
+assembly name hashes.  All the blobs are assigned a unique ID, with
+the blob having ID equal to `0` being the [Index blob](#index-blob)
+
+Assemblies are stored as adjacent byte streams:
+
+ - **Image data**
+   Required to be present for all assemblies, contains the actual
+   assembly PE image.
+ - **Debug data**
+   Optional. Contains the assembly's PDB or MDB debug data.
+ - **Config data**
+   Optional. Contains the assembly's .config file. Config data
+   **must** be terminated with a `NUL` character (`0`), this is to
+   make runtime code slightly more efficient.
+
+All the structures described here are defined in the
+[`xamarin-app.hh`](../../src/monodroid/jni/xamarin-app.hh) file.
+Should there be any difference between this document and the
+structures in the header file, the information from the header is the
+one that should be trusted.
+
+## Common header
+
+All kinds of blobs share the following header format:
+
+    struct BundledAssemblyBlobHeader
+    {
+        uint32_t magic;
+        uint32_t version;
+        uint32_t local_entry_count;
+        uint32_t global_entry_count;
+        uint32_t blob_id;
+    ;
+
+Individual fields have the following meanings:
+
+ - `magic`: has the value of 0x41424158 (`XABA`)
+ - `version`: a value increased every time blob format changes.
+ - `local_entry_count`: number of assemblies stored in this blob (also
+   the number of entries in the assembly descriptor table, see below)
+ - `global_entry_count`: number of entries in the index blob's (see
```

Comment:
> I don't understand.
> 
> I think we need more nouns. :-/
> 
> Thus, a suggestion/clarification:
> 
> There are the following blob _files_:
> 
>     * App Architecture-Independent Assembly [& related] Blob (`assemblies.blob` in main app)
> 
>     * App Architecture-Specific Assembly Blobs (`assemblies.[ARCH].blob` in main app)
> 
>     * Feature Architecture-Independent Assembly [& related] Blob (zero or more) (`assemblies.blob` in feature)
> 
>     * Feature Architecture-Specific Assembly Blobs (zero or more) (`assemblies.[ARCH].blob` in feature)
> 
> 
> _Each_ of those blob files has an internal structure.

Which is identical for all the blobs, with the only difference being the global hash index which is **not** part of the general blob header.

> 
> We thus return to `BundledAssemblyBlobHeader::global_entry_count`: _what is it_, and for _where_?

A count of all the entries in all the blobs in the application together - a **global** count of assemblies, versus a **local** count in a single blob file.

> 
> For the App Architecture-Independent Blob, is it the count of _all_ of the above (all App + Feature blobs)? Just the "current" blob". Just the _App_ blobs? Other?

It's a global count for the entire app - all the blobs summed up. It's essentially the size of the pre-generated table in `libxamarin-app.so`

> 
> The "all the other blobs store `0` in this field" sentence fragment (below) is not helping.

Not sure what you mean, it's a format specification and it clearly states that this field is 0 for all the blobs except the index one.

> 
> I _imagine_ that it may be _possible_ to update a Feature without updating the corresponding App -- and, if supported, would be _beneficial_, as it means less data needs to be sent to devices -- so I _hope_ that the App Arch-Independent Blob's `global_entry_count` _doesn't_ include counts for _all related Feature blobs_.

I doubt it will be possible to update a Feature without updating the app.  With the AAB format being the required distribution format for the Play Store, I don't think it's possible to update the app piecemeal - how would it be done? By uploading a "mini-AAB" with just the feature update in it? I rather think the apps will be uploaded as a single AAB from which feature APKs will be created by Google, on the Play Store side. So the app will be rebuilt entirely and there's no issue with this global count.